packages/taquito/src/operations/types.ts

```diff
@@ -563,6 +566,53 @@ export interface FailingNoopParams {
   basedOnBlock: BlockIdentifier;
 }
 
+export interface InternalStakingParams {
+  source: string;
+  amount: number;
+  fee?: number;
+  gasLimit?: number;
+  storageLimit?: number;
+  mutez?: boolean;
+  parameter: TransactionOperationParameter;
+}
+
+type PartialBy<T, K extends keyof T> = Omit<T, K> & Partial<Pick<T, K>>;
+export type StakingParams = PartialBy<InternalStakingParams, 'source'>;
+export type StakeParams = Omit<StakingParams, 'parameter'>;
+export type UnstakeParams = Omit<StakingParams, 'parameter'>;
+export type FinalizeUnstakeParams = Omit<StakingParams, 'parameter' | `amount`>;
+
+/**
+ *
+ * @throws {@link InvalidAmountError}
+ * @throws {@link InvalidAddressError}
+ * @throws {@link InvalidStakingSource}
+ */
+export const validateStakingParams = (
+  params: Omit<PartialBy<StakingParams, 'amount'>, 'parameter'>,
+  entrypoint: StakingEntrypoint
+) => {
+  if (entrypoint !== 'finalize_unstake') {
+    if (!params.amount || params.amount <= 0) {
+      throw new InvalidAmountError(
+        params.amount?.toString() ?? '',
+        `The 'amount' field in ${entrypoint} should be a positive number.`
+      );
+    }
+  }
+  if (params.source !== undefined) {
+    const sourceValidation = validateAddress(params.source);
```

Comment:
validateKeyHash will validate only implicit line608 can be removed

Reply:
I did that, but then realized that I prefer to have different errors in case the address is in wrong format, vs when the address has an unsupported prefix.
Undoing the change I made in response to this comment now.

Reply:
in general all transaction op can't have kt1 as source, `packages/taquito/src/prepare/prepare-provider.ts:239` have addressed that not necessary for extra InvalidStakingSource

Reply:
No problem, I can remove this check.